pallets/xcmp-queue/src/lib.rs

```diff
@@ -941,84 +510,173 @@ impl<T: Config> Pallet<T> {
+				log::error!("Cannot suspend channel from sibling {:?}: {:?}; further messages may be dropped.", para, err);
+			} else if let Err(err) = suspended_channels.try_insert(para) {
+				log::error!("Too many channels suspended; cannot suspend sibling {:?}: {:?}; further messages may be dropped.", para, err);
+			} else {
+				<InboundXcmpSuspended<T>>::put(suspended_channels);
+			}
+		}
+	}
+}
+
+impl<T: Config> QueuePausedQuery<ParaId> for Pallet<T> {
+	fn is_paused(para: &ParaId) -> bool {
+		if !QueueSuspended::<T>::get() {
+			return false
+		}
+
+		let sender_origin = T::ControllerOriginConverter::convert_origin(
+			(Parent, Parachain((*para).into())),
+			OriginKind::Superuser,
+		);
+		let is_controller =
+			sender_origin.map_or(false, |origin| T::ControllerOrigin::try_origin(origin).is_ok());
+
+		!is_controller
+	}
 }
 
 impl<T: Config> XcmpMessageHandler for Pallet<T> {
 	fn handle_xcmp_messages<'a, I: Iterator<Item = (ParaId, RelayBlockNumber, &'a [u8])>>(
 		iter: I,
 		max_weight: Weight,
 	) -> Weight {
-		let mut status = <InboundXcmpStatus<T>>::get();
+		let mut meter = WeightMeter::from_limit(max_weight);
 
-		let QueueConfigData { suspend_threshold, drop_threshold, .. } = <QueueConfig<T>>::get();
-
-		for (sender, sent_at, data) in iter {
-			// Figure out the message format.
-			let mut data_ref = data;
-			let format = match XcmpMessageFormat::decode_with_depth_limit(
-				MAX_XCM_DECODE_DEPTH,
-				&mut data_ref,
-			) {
+		for (sender, _sent_at, mut data) in iter {
+			let format = match XcmpMessageFormat::decode(&mut data) {
 				Ok(f) => f,
 				Err(_) => {
-					debug_assert!(false, "Unknown XCMP message format. Silently dropping message");
+					defensive!("Unknown XCMP message format. Message silently dropped.");
 					continue
 				},
 			};
-			if format == XcmpMessageFormat::Signals {
-				while !data_ref.is_empty() {
-					use ChannelSignal::*;
-					match ChannelSignal::decode(&mut data_ref) {
-						Ok(Suspend) => Self::suspend_channel(sender),
-						Ok(Resume) => Self::resume_channel(sender),
-						Err(_) => break,
-					}
-				}
-			} else {
-				// Record the fact we received it.
-				match status.binary_search_by_key(&sender, |item| item.sender) {
-					Ok(i) => {
-						let count = status[i].message_metadata.len();
-						if count as u32 >= suspend_threshold && status[i].state == InboundState::Ok
+
+			match format {
+				XcmpMessageFormat::Signals =>
+					while !data.is_empty() {
+						if meter
+							.try_consume(
+								T::WeightInfo::suspend_channel()
+									.max(T::WeightInfo::resume_channel()),
+							)
+							.is_err()
 						{
-							status[i].state = InboundState::Suspended;
-							let r = Self::send_signal(sender, ChannelSignal::Suspend);
-							if r.is_err() {
-								log::warn!(
-									"Attempt to suspend channel failed. Messages may be dropped."
-								);
-							}
+							defensive!("Not enough weight to process signals - dropping.");
```

Comment:
Hm its difficult... with the current logic it will try to re-send the suspend signal when the next message on a suspended channel comes in. This goes on until the `drop_threshold` is reached, so i think with proper configuration this should be fine.  
I am always very hesistant to ignore the weight limit, its a hard limit after all.